contracts/Redemption.vy

```diff
@@ -0,0 +1,502 @@
+
+
+# @dev Emitted when the ownership is transferred
+# from `previous_owner` to `new_owner`.
+event OwnershipTransferred:
+    previous_owner: indexed(address)
+    new_owner: indexed(address)
+
+event SetPayee:
+    payee: indexed(address)
+
+
+MASK: constant(uint256) = 2**64 - 1
+
+# powers of 10
+E3: constant(int256)               = 1_000
+E6: constant(int256)               = E3 * E3
+E9: constant(int256)               = E3 * E6
+E12: constant(int256)              = E3 * E9
+E15: constant(int256)              = E3 * E12
+E17: constant(int256)              = 100 * E15
+E18: constant(int256)              = E3 * E15
+E20: constant(int256)              = 100 * E18
+MIN_NAT_EXP: constant(int256)      = -41 * E18
+MAX_NAT_EXP: constant(int256)      = 130 * E18
+
+# x_n = 2^(7-n), a_n = exp(x_n)
+# in 20 decimals for n >= 2
+X0: constant(int256)  = 128 * E18 # 18 decimals
+A0: constant(int256)  = 38_877_084_059_945_950_922_200 * E15 * E18 # no decimals
+X1: constant(int256)  = X0 / 2 # 18 decimals
+A1: constant(int256)  = 6_235_149_080_811_616_882_910 * E6 # no decimals
+X2: constant(int256)  = X1 * 100 / 2
+A2: constant(int256)  = 7_896_296_018_268_069_516_100 * E12
+X3: constant(int256)  = X2 / 2
+A3: constant(int256)  = 888_611_052_050_787_263_676 * E6
+X4: constant(int256)  = X3 / 2
+A4: constant(int256)  = 298_095_798_704_172_827_474 * E3
+X5: constant(int256)  = X4 / 2
+A5: constant(int256)  = 5_459_815_003_314_423_907_810
+X6: constant(int256)  = X5 / 2
+A6: constant(int256)  = 738_905_609_893_065_022_723
+X7: constant(int256)  = X6 / 2
+A7: constant(int256)  = 271_828_182_845_904_523_536
+X8: constant(int256)  = X7 / 2
+A8: constant(int256)  = 164_872_127_070_012_814_685
+X9: constant(int256)  = X8 / 2
+A9: constant(int256)  = 128_402_541_668_774_148_407
+X10: constant(int256) = X9 / 2
+A10: constant(int256) = 11_331_4845_306_682_631_683
+X11: constant(int256) = X10 / 2
+A11: constant(int256) = 1_064_49_445_891_785_942_956
+
+@external
+def __init__(
+    yfi: address, d_yfi: address, ve_yfi: address, owner: address, 
+    price_feed: address, curve_pool: address, scaling_factor: uint256,
+):
+    assert scaling_factor >= UNIT and scaling_factor <= 12 * UNIT
+    YFI = ERC20(yfi)
+    DYFI = IDYFI(d_yfi)
+    VEYFI = ERC20(ve_yfi)
+    PRICE_FEED = AggregatorV3Interface(price_feed)
+    CURVE_POOL = CurvePoolInterface(curve_pool)
+    self._transfer_ownership(owner)
+    self.payee = owner
+    self.packed_scaling_factor = shift(scaling_factor, 128) | shift(scaling_factor, 192)
+
+
+@payable
+@external
+def redeem(amount: uint256, recipient: address = msg.sender) -> uint256:
+    """
+    @notice Redeem your dYFI for YFI using ETH.
+    @dev Redemption tolerates a 0.3% negative or positive slippage.
+    @param amount amount of dYFI to spend
+    @param recipient of the exercised YFI
+    """
+    self._check_killed()
+    assert YFI.balanceOf(self) >= amount, "not enough YFI"
+    eth_required: uint256 = self._eth_required(amount)
```

Comment:
it could be good for additional safety to store the eth_required in a var and compare with latest read for acceptable deviation.

```
if (abs(eth_required - last_eth_required) / last_eth_required > DEVIATION_TOLERANCE) {
   // means something is way off on the price feed
} else {
   last_eth_required = eth_required
}
```

this is probably optional but happy to discuss